api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -543,6 +568,124 @@ public int hashCode() {
     }
   }
 
+  public static class GeometryType extends PrimitiveType {
+
+    private final String crs;
+
+    private GeometryType(String crs) {
+      this.crs = (crs == null || crs.isEmpty()) ? null : crs;
+    }
+
+    public static GeometryType crs84() {
+      return new GeometryType(null);
+    }
+
+    public static GeometryType of(String crs) {
+      return new GeometryType(crs);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOMETRY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeometryType)) {
+        return false;
+      }
+
+      GeometryType that = (GeometryType) o;
+      return Objects.equals(crs, that.crs);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeometryType.class, crs);
+    }
+
+    @Override
+    public String toString() {
+      if (Strings.isNullOrEmpty(crs)) {
+        return "geometry";
+      }
+
+      return String.format("geometry(%s)", crs);
+    }
+  }
+
+  public static class GeographyType extends PrimitiveType {
+
+    private final String crs;
+    private final EdgeAlgorithm algorithm;
+
+    private GeographyType(String crs, EdgeAlgorithm algorithm) {
+      this.crs = (crs == null || crs.isEmpty()) ? null : crs;
+      this.algorithm = algorithm;
+    }
+
+    public static GeographyType crs84() {
+      return new GeographyType(null, null);
+    }
+
+    public static GeographyType of(String crs) {
+      return new GeographyType(crs, null);
+    }
+
+    public static GeographyType of(String crs, EdgeAlgorithm algorithm) {
+      return new GeographyType(crs, algorithm);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOGRAPHY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    public EdgeAlgorithm algorithm() {
+      return algorithm;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeographyType)) {
+        return false;
+      }
+
+      GeographyType that = (GeographyType) o;
+      return Objects.equals(crs, that.crs) && Objects.equals(algorithm, that.algorithm);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeographyType.class, crs, algorithm);
+    }
+
+    @Override
+    public String toString() {
+      if (algorithm != null) {
+        return String.format(
+            "geography(%s, %s)",
+            crs != null ? crs : "", algorithm.name().toLowerCase(Locale.ENGLISH));
```

Comment:
I think it would be fine to move the `toLowerCase` into the enum. We just don't need to parameterize the enum with a constant that is the lower case name.

Reply:
Implemented `toString` in the enum to perform `toLowerCase`.